ykrt/src/location.rs

```diff
@@ -97,88 +29,408 @@ impl State {
+        let mut ls = self.load(Ordering::Relaxed);
+        loop {
+            if ls.is_locked() {
+                return Err(());
+            }
+            let new_ls = ls.with_lock();
+            match self.compare_exchange_weak(ls, new_ls, Ordering::Acquire, Ordering::Relaxed) {
+                Ok(_) => return Ok(new_ls),
+                Err(x) => ls = x,
+            }
+        }
     }
 }
 
 impl<I> Drop for Location<I> {
     fn drop(&mut self) {
-        let lp = *self.state.get_mut();
-        if lp & PHASE_TAG == PHASE_TRACING {
-            unsafe { Arc::from_raw((lp & !PHASE_TAG) as *mut u8) };
+        let ls = self.load(Ordering::Relaxed);
+        if !ls.is_counting() {
+            debug_assert!(!ls.is_locked());
+            let hr = ls.hot_location();
+            unsafe {
+                Box::from_raw(hr);
+            }
+        }
+    }
+}
+
+#[cfg(target_pointer_width = "64")]
+pub(crate) const STATE_TAG: usize = 0b111; // All of the other tag data must fit in this.
+#[cfg(target_pointer_width = "64")]
+pub(crate) const STATE_NUM_BITS: usize = 3;
+
+pub(crate) const STATE_IS_LOCKED: usize = 0b001;
+pub(crate) const STATE_IS_PARKED: usize = 0b010;
+pub(crate) const STATE_IS_COUNTING: usize = 0b100;
+
+pub(crate) const TOKEN_NORMAL: UnparkToken = UnparkToken(0);
+pub(crate) const TOKEN_HANDOFF: UnparkToken = UnparkToken(1);
+
+#[derive(PartialEq, Eq, Debug)]
+pub(crate) struct State<I> {
+    pub(crate) x: usize,
+    marker: PhantomData<I>,
+}
+
+impl<I> Copy for State<I> {}
+
+impl<I> Clone for State<I> {
+    fn clone(&self) -> State<I> {
+        *self
+    }
+}
+
+impl<I> State<I> {
+    /// Return a new `State` in the counting phase with a count 0.
+    pub(crate) fn new() -> Self {
+        State {
+            x: STATE_IS_COUNTING,
+            marker: PhantomData,
+        }
+    }
+
+    fn from_usize(x: usize) -> Self {
+        State {
+            x,
+            marker: PhantomData,
+        }
+    }
+
+    pub(crate) fn is_locked(&self) -> bool {
+        self.x & STATE_IS_LOCKED != 0
+    }
+
+    pub(crate) fn is_parked(&self) -> bool {
+        self.x & STATE_IS_PARKED != 0
+    }
+
+    /// Is the Location in the counting or a non-counting state?
+    pub(crate) fn is_counting(&self) -> bool {
+        self.x & STATE_IS_COUNTING != 0
+    }
+
+    /// If, and only if, the Location is in the counting state, return the current count.
+    pub(crate) fn count(&self) -> usize {
+        debug_assert!(self.is_counting());
+        debug_assert!(!self.is_locked());
+        self.x >> STATE_NUM_BITS
+    }
+
+    /// If this `State` is not counting, return its `HotLocation`. It is undefined behaviour to
+    /// call this function if this `State` is in the counting phase and/or if this `State` is not
+    /// locked.
+    pub(crate) fn hot_location(&self) -> &mut HotLocation<I> {
+        debug_assert!(!self.is_counting());
+        //debug_assert!(self.is_locked());
```

Comment:
This is unsound without these assertions being unconditional, so it need to be unsafe.

Reply:
Agreed. Fixed in e2a418384ea6ccda3d9bc3fe6f38103963205989.